ortb/clone.go

```diff
@@ -0,0 +1,271 @@
+package ortb
+
+import (
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/util/ptrutil"
+	"github.com/prebid/prebid-server/util/sliceutil"
+)
+
+func CloneApp(s *openrtb2.App) *openrtb2.App {
+	if s == nil {
+		return nil
+	}
+
+	// Shallow Copy (Value Fields)
+	c := *s
+
+	// Deep Copy (Pointers)
+	c.Cat = sliceutil.Clone(s.Cat)
+	c.SectionCat = sliceutil.Clone(s.SectionCat)
+	c.PageCat = sliceutil.Clone(s.PageCat)
+	c.Publisher = ClonePublisher(s.Publisher)
+	c.Content = CloneContent(s.Content)
+	c.KwArray = sliceutil.Clone(s.KwArray)
+	c.Ext = sliceutil.Clone(s.Ext)
+
+	return &c
+}
+
+func ClonePublisher(s *openrtb2.Publisher) *openrtb2.Publisher {
+	if s == nil {
+		return nil
+	}
+
+	// Shallow Copy (Value Fields)
+	c := *s
+
+	// Deep Copy (Pointers)
+	c.Cat = sliceutil.Clone(s.Cat)
+	c.Ext = sliceutil.Clone(s.Ext)
+
+	return &c
+}
+
+func CloneContent(s *openrtb2.Content) *openrtb2.Content {
+	if s == nil {
+		return nil
+	}
+
+	// Shallow Copy (Value Fields)
+	c := *s
+
+	// Deep Copy (Pointers)
+	c.Producer = CloneProducer(s.Producer)
+	c.Cat = sliceutil.Clone(s.Cat)
+	c.ProdQ = ptrutil.Clone(s.ProdQ)
+	c.VideoQuality = ptrutil.Clone(s.VideoQuality)
+	c.KwArray = sliceutil.Clone(s.KwArray)
+	c.Data = CloneDataSlice(s.Data)
+	c.Network = CloneNetwork(s.Network)
+	c.Channel = CloneChannel(s.Channel)
+	c.Ext = sliceutil.Clone(s.Ext)
+
+	return &c
+}
+
+func CloneProducer(s *openrtb2.Producer) *openrtb2.Producer {
+	if s == nil {
+		return nil
+	}
+
+	// Shallow Copy (Value Fields)
+	c := *s
+
+	// Deep Copy (Pointers)
+	c.Cat = sliceutil.Clone(s.Cat)
+	c.Ext = sliceutil.Clone(s.Ext)
+
+	return &c
+}
+
+func CloneDataSlice(s []openrtb2.Data) []openrtb2.Data {
+	if s == nil {
+		return nil
+	}
+
+	c := make([]openrtb2.Data, len(s))
+	for i, d := range s {
+		c[i] = CloneData(d)
+	}
+
+	return c
+}
+
+func CloneData(s openrtb2.Data) openrtb2.Data {
+	// Shallow Copy (Value Fields)
+	// - Already copied from argument pass by value.
```

Comment:
Nitpick: I feel like this comment could be combined into one, something like `// Shallow copy already occurred from argument passed by value`. But if you feel otherwise, I'm happy to leave as is.

Reply:
Updated all occurrences.